src/topic.rs

```diff
@@ -0,0 +1,56 @@
+use std::fmt::{Debug, Display, Formatter};
+use std::sync::Arc;
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct Topic(Arc<String>);
+
+impl AsRef<String> for Topic {
+    fn as_ref(&self) -> &String {
+        self.0.as_ref()
+    }
+}
+impl Debug for Topic {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        if f.alternate() {
+            write!(f, "{:#?}", self.0)
+        } else {
+            write!(f, "Topic({:#?})", self.0)
+        }
+    }
+}
+impl From<&str> for Topic {
+    fn from(s: &str) -> Self {
+        s.to_string().into()
+    }
+}
+impl From<String> for Topic {
+    fn from(string: String) -> Self {
+        Topic(Arc::new(string))
+    }
+}
+impl From<Topic> for Arc<String> {
+    fn from(topic: Topic) -> Self {
+        topic.0.clone()
+    }
+}
+impl From<Topic> for serde_json::Value {
```

Comment:
Better to use Serialize/Deserialize

Reply:
It can't be derived because of the `Arc<Sring>`:

```
error[E0277]: the trait bound `std::sync::Arc<std::string::String>: Serialize` is not satisfied
   --> src/topic.rs:5:55
    |
5   | #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
    |                                                       ^^^^^^^^^ the trait `Serialize` is not implemented for `std::sync::Arc<std::string::String>`
6   | #[repr(transparent)]
7   | pub struct Topic(Arc<String>);
    |                  ----------- required by a bound introduced by this call
    |
    = help: the following other types implement trait `Serialize`:
              &'a T
              &'a mut T
              ()
              (T0, T1)
              (T0, T1, T2)
              (T0, T1, T2, T3)
              (T0, T1, T2, T3, T4)
              (T0, T1, T2, T3, T4, T5)
            and 128 others
note: required by a bound in `serialize_newtype_struct`
   --> /Users/kronic.deth/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.145/src/ser/mod.rs:904:12
    |
904 |         T: Serialize;
    |            ^^^^^^^^^ required by this bound in `Serializer::serialize_newtype_struct`

```

Reply:
Serde has the `rc` feature flag for enabling this or you can use flexstr with the `serde` feature flag. The latter is more efficient because there's only a single heap allocation at most

Reply:
✅